release 3.5.0 (#6131)
open-formulieren/open-forms: CHANGELOG.rst

        `latest <https://open-forms.readthedocs.io/en/latest/changelog.html>`_ docs
        version.

> 3.5.0 (2026-04-13)

Reviewer: ```suggestion
3.5.0 (2026-04-15)
```
I don't think we'll make it today 😬

==================

Open Forms 3.5.0 is a feature release.

This contains the changes from the alpha releases and fixes applied until the stable version.
BEFORE upgrading to 3.5.0, please read the release notes carefully and review the following
instructions.

.. note::

    Starting with Open Forms 3.5 the new renderer is enabled by default when a new
    form is created. If you encounter problems, you can revert back to the old renderer

>     in the form settings.

Reviewer: This is mentioned on the detailed upgrade notes (https://open-forms.readthedocs.io/en/latest/installation/upgrade-350.html) so I think we can omit it from here or keep it as a normal feature entry that the defaults have been updated.

Author: The note is removed. Added a feature entry in "Performance" section

Upgrade procedure
-----------------

To upgrade to 3.5, please:

> * ⚠️ Ensure you are currently on Open Forms 3.4.x or newer.

Reviewer: What's the reason to require 3.4.x? In the code I see 3.3.1 as minimum version:

```py
UPGRADE_CHECK_PATHS: UpgradePaths = {
    "3.4.0": UpgradeCheck(
        VersionRange(minimum="3.3.1"),
    ),
}
```
(there probably needs to be an entry for 3.5.0 too)

Author: Discussed offline. Changed to 3.3.1 and updated `UPGRADE_CHECK_PATHS` entry

  prepare accordingly.
* ⚠️ Ensure that your Helm charts don't use the ``bin/check_celery_worker_liveness.py``
  script, since it has been removed. You can read about the new health check mechanisms
  here: :ref:`installation_health_checks`.
* ⚠️ Logic rules using the "disable next" action may need extra attention.

  .. warning::

      In preparation for upcoming logic performance improvements, we changed the
      "disable next" logic action. The form designers now have to specify on which step
      they want to execute this action. We included a migration to automatically
      populate this field for all existing actions. Unfortunately, in some cases we
      cannot guarantee that the assigned step is correct, i.e. no change in behaviour
      is observed when filling out a form. Therefore, form designers/maintainers are
      asked to check these cases manually. The script below outputs a list of logic
      rules (with their corresponding form) that need manual attention. Use the flag
      ``--show-all`` to output a complete list of all affected rules.

      .. code-block:: bash

          # in the container via ``docker exec`` or ``kubectl exec``:

>           python /app/bin/check_disable_next_logic_action.py

Reviewer: This is no longer "in preparation for upcoming...", but is now the actual case. The script reports which forms will have been updated and need to be checked.

IIRC the migration affects all forms, even if you don't opt into new logic evaluation (because you can only do that after upgrading). @viktorvanwijk can you confirm? & do we perhaps need to backport this check script to 3.4.x so they can prepare accordingly?

Author: removed "in preparation for upcoming..."

> We'll work on bridging this gap, possibly through an upgrade to Django 6.0.

Reviewer: I think we need to cross-check the already announced deprecations in earlier releases against #6164 to make sure all deprecations have been mentioned. I've the feeling there are a bunch now that haven't been officially deprecated yet.

Reviewer: ## Missing deprecations

* Textfield address autofill support. Needs to be done through addressNL in the future.
* Legacy logic evaluation.
* Documenttype configuration by URL in the file component (registration tab). Will move
  to the registration plugin options.
* Elastic-APM will be removed.
* (ZGW APIs/StUF-ZDS) registration: defaults will change (public reference generation?),
  KVK branch number will be taken from auth context before form component if both are set.
* `authentication_success` signal will be deleted
* Formio registry `build_serializer_field` fallback will be removed.
* DigiD/eHerkenning via OIDC will switch from Lax to Strict mode by default.
* Haal Centraal BRP Personen 1.3 support will be removed.
* The behaviour of `clearOnHide` will change: instead of using the `defaultValue` for a
  cleared value, the whole variable will be removed/absent. This will affect existing
  logic rules.
* The public interfaces for plugins/extensions will become semi-private API allowing us
  to make potentially breaking changes without making a new major version of OF. 
  Extensions do not seem to be created commonly by third parties and may be removed in
  the future.

## Already mentioned in earlier releases

### 3.4.0

* Legacy renderer (formio.js)
* Styling shims for backwards compatibility
* Ogone payment plugin
* JCC SOAP service
* UMD bundle (SDK)
* Bundles moved (SDK)
* Upload component design tokens, including loading indicator and modal.

### 3.2.0

* Camunda registration backend deprecated

### 3.0.0

* Legacy OIDC callback endpoints
* ZGW APIs URL-based configuration
* Objects API URL-based configuration

Author: I haven't included into changelog (since they seem to be for developers only):
* `authentication_success` signal will be deleted
* Formio registry`build_serializer_field` fallback will be removed.

Other deprecations are added to the best of my ability

Major features
--------------

**⚡️ Performance**

We've done a lot of work to make the interaction of users with Open Forms smoother and faster.
One of the main culprits of micro-delays is a logic check that ran every time when the user
changes the input data, re-evaluating all logic rules each time.

> We've done an extensive refactoring of the logic rules to eliminate the unnecessary checks:

Reviewer: ```suggestion
We have extensively reworked our logic engine to eliminate or reduce network traffic:
```

- only the logic rules relevant to the current form step are now executed,
- logic rules are ordered to ensure the correct order of the chain of rules.

> - cycles in logic rules are detected,

Reviewer: ```suggestion
- rules that would lead to infinite loops are now prevented,
```

> - some of the checks are moved to the frontend.

Reviewer: ```suggestion
- the most commonly used logic rule patterns are now evaluated on the frontend.
```

Both the old and new logic evaluations are available, and you can enable the new one on
an individual form basis.

> Note: executing logic checks in the fronend is only available for the new renderer.

Reviewer: ```suggestion
Note: executing logic checks in the frontend is only available for the new renderer.
```

**🔍 Monitoring and logging**

We've put a lot of work to make Open Forms easier to monitor and troubleshoot:

- All services now have built-in health checks, so problems can be detected earlier.
- We fully support Open telemetry tracing across the entire application, covering DB and redis queries,
  requests to external services, application-specic traces etc., which improves visibility of the whole submission flow.

> - We support inking to traces from logs in Grafana.

Reviewer: ```suggestion
- Traces and logs are now correlated, making it possible to link from one to the other in dashboards.
```

- Errors captured in Sentry can now be correlated with traces and logs, which can help in investigating problems.
- All logs are now structured.
- Audit logs are improved: they are now sent to both the database and the system output of containers.

These improvements could help with faster troubleshooting and better insight into how Open Forms is performing

> in production.

Reviewer: ```suggestion
These improvements enable faster troubleshooting and better insight into how Open Forms is performing
in production.
```

**🔢 Stable public references**

Previously, public references were generated randomly, which could lead to duplicate reference
numbers when old submissions were pruned. References are now generated deterministically guaranteeing unique numbers
even when old submissions are deleted. Administrators can configure a custom template for the reference format.
Additionally, for the ZGW APIs registration plugin, you can now choose between the generated case numbers and the references

> generated by Open Forms.

Reviewer: ```suggestion
**🔢 Stable public references**

Previously, public references were generated randomly, which (theoretically) could lead to duplicate reference
numbers when old submissions were pruned. References are now generated deterministically guaranteeing unique numbers
even when old submissions are deleted. Administrators can configure a custom template for the reference format.
Additionally, for the ZGW APIs registration plugin, you can now choose between the generated case numbers and the references
generated by Open Forms.
```

Please move this above logging and monitoring as well :-)

Author: Changed and moved above "monitoring"

**✉️ Support for sending mail via Office365 Graph API**

Microsoft is deprecating SMTP Basic authentication and `plans <https://techcommunity.microsoft.com/blog/exchange/updated-exchange-online-smtp-auth-basic-authentication-deprecation-timeline/4489835>`_
to remove it entirely in 2027. To keep email delivery working for Office365 users, Open Forms now supports
sending emails via the Microsoft Graph API, thanks to a contribution by `Delta10 <https://www.delta10.nl/>`_.

New features
------------

* Performance:

  .. note::

      We improved the performance of logic rule evaluation in this release. These changes
      are enabled by default. You can turn them off by unchecking "Enable new logic rule

>       evaluation" feature flag.

Reviewer: ```suggestion
      We improved the performance of logic rule evaluation in this release. These changes
      are enabled by default for newly created forms. You can turn them off by unchecking 
      the "Enable new logic rule evaluation" feature flag.
```

      On form save, logic rules are validated and re-ordered. If you modify individual
      logic rules (in the admin interface), this can result in out-of-sync analysis. You

>       should avoid doing this.

Reviewer: ```suggestion
      With the new logic evaluation,  logic rules are validated and re-ordered on form save. If you modify individual
      logic rules (in the admin interface), this can result in out-of-sync analysis. You
      should avoid doing this.
```

    variables are used in logic rules.
  - [:backend:`2409`] Open Forms now detects cycles in logic rules when one rule depends on the
    outcome of one or more other rules. Cycles like ``ruleA -> ruleB -> ruleC -> ruleA`` are now caught
    and validation errors are shown.
  - [:backend:`2409`] Logic rules are now automatically re-ordered when you opt in to the new logic
    evaluation, and assigned to the relevant form steps so that only relevant logic rules are evaluated
    on a particular form step.
  - [:backend:`5962`] Implemented partial logic evaluation of the logic rules.
  - [:backend:`6038`] JsonLogic expressions can now be compiled with data-type
    information for the frontend.
  - [:backend:`6037`] Set up shared jsonLogic test suite (between backend and frontend).
  - [:backend:`6072`] Implement "require backend" analysis for logic rules, so for each logic action
    we know now if the backend evaluation is needed.
  - Added logic rules to the submission step serializer.
  - [:backend:`6083`] Added partial evaluation of the variable action before sending logic
    rules to frontend.

>   - [:backend:`6109`] Removed ``completed`` and ``is_applicable`` fields from the submission step.

Reviewer: ```suggestion
  - [:backend:`6109`] Removed obsolete ``completed`` and ``is_applicable`` fields from the submission step.
```

Otherwise it sounds like a breaking change, but this is a private endpoint not subject to SemVer

  - [:backend:`6038`] Ensure "set-registration-backend" actions are not included in frontend logic rules.

>   - [:backend:`6099`] Executed step (not) applicable action on all steps.

Reviewer: This entry can be removed, it was a tweak of the actual features above.

Author: Removed

>   - [:backend:`6018`, :backend:`6129`] Removed ``SubmissionStep.data`` property.

Reviewer: ```suggestion
  - [:backend:`6018`, :backend:`6129`] Removed the obsolete ``SubmissionStep.data`` property.
```

  - [:backend:`6143`] Added documentation about accessing the variables state from the submission for
    developers.
  - Added metric for forms with new logic evaluation enabled.
  - [:backend:`6018`] Cleaned up variable names and updated comments and docstrings related to
    component visibility implementation.
  - [:backend:`6166`] Fixed list comparisons in JsonLogic expressions.

* New Admin UI:

  .. note::

      After investing a lot of effort in designing the new admin interface, we started the implementation phase.

>       Unfortunately due to the budjet issues this project is on hold.

Reviewer: ```suggestion
      Unfortunately due to the budget issues this project is on hold.
```

  - [:backend:`5952`] Added initial form endpoint.
  - [:backend:`5955`] Added support for confirmation templates/options in the new endpoint.
  - [:backend:`6080`] Added ``uuid`` to the ``/api/v2/themes`` endpoint response.
  - [:backend:`5952`] Added the ``steps`` field to the new v3 ``/api/v3/forms/{uuid}`` endpoint.
  - [:backend:`5956`] Added the ``registration_backends`` field to the new v3 ``/api/v3/forms/{uuid}`` endpoint.
  - [:backend:`5957`] Added payment related fields to the new v3 ``/api/v3/forms/{uuid}`` endpoint.
  - [:backend:`6119`] Supported zero steps for appointment forms in the new v3 endpoint.

* Appointments (JCC):

  - [:backend:`5690`] The additional product information (HTML) is now exposed to the SDK.
  - [:backend:`5696`] Implemented the remaining API calls for the appointment flow. We now fully support REST API,

>     since SOAP is decommissioned by JCC.

Reviewer: ```suggestion
  - [:backend:`5696`] Implemented the remaining API calls for the appointment flow. We now fully support the REST API,
    since SOAP is decommissioned by JCC.
```

  - [:backend:`5820`] You can now customize the literals and translations used in the contact details components.

>   - [:backend:`5691`] Added a limit for the number of products in appointments.

Reviewer: ```suggestion
  - [:backend:`5691`] We now expose the limit for the number of products in appointments to the frontend.
```

* Monitoring and logging:

  - [:backend:`5287`] All containers now have low-level health check mechanisms.

>   - [:backend:`5450`] Distributed tracing is supported. The following component traces are now available:

Reviewer: ```suggestion
  - [:backend:`5450`] We added support for distributed tracing with Open Telemetry. The following component traces are now available:
```

    - Postgres queries
    - Redis queries
    - Celery tasks
    - Outgoing requests to external services
    - Application-specific traces

  - [:backend:`5358`] Added span/trace IDs to log events for correlation and supported linking to traces from logs

>     in Grafana.

Reviewer: ```suggestion
  - [:backend:`5358`] Added span/trace IDs to log events for correlation.
```
which dashboard/storage users use is up to them, Grafana is mostly for ourselves

  - Captured ``TraceID`` and ``SpanID`` from OTel for Sentry events to allow correlation between
    Sentry events and traces/logs in Grafana.
  - [:backend:`5351`, :backend:`6149`] Refactored ``logevent`` module to structlog adapters: audit logs are now
    sent to both container stdout and database.
  - [:backend:`6169`] Improved intergration of outgoing requests logs with structlog. All logs are now structured.

>   - Logged file step attachment additions, which can be used in the dashboards.

Reviewer: ```suggestion
  - File step attachment additions are now logged, which can produce exact insights into typical file upload amounts per form.
```

* [:backend:`5753`] Public reference generation is now deterministic and guaranteed to never generate
  duplicated reference numbers, even if old submissions have been deleted. Administrators can configure
  a custom template to use. For the ZGW APIs registration plugin, you can now choose between the

>   generated case numbers and the references generated by Open Forms.

Reviewer: ```suggestion
* [:backend:`5753`] The generation of the public reference is now deterministic and guaranteed to never produce
  duplicated reference numbers, even if old submissions have been deleted. Administrators can configure
  a custom template to use. For the ZGW APIs registration plugin, you can now choose between the
  case numbers obtained from the Zaken API or the references generated by Open Forms.
```

* [:backend:`5972`] Added support for sending e-mails via the Office365 Graph API.
* [:backend:`5319`] The "Organization login via OpenID Connect" can now be hidden by

>   default. Use the ``authVisible=all`` query parameter to make it visible in the SDK.

Reviewer: ```suggestion
* [:backend:`5319`] The "Organization login via OpenID Connect" can now be hidden by
  default. Use the ``authVisible=all`` query parameter to make it visible in the public frontend.
```

> * [:backend:`5820`] Added endpoint to retrieve the overridden frontend translations.

Reviewer: ```suggestion
* [:backend:`5820`] Added endpoint to retrieve overridden frontend translations.
```

* [:backend:`5553`] When starting a form from organization login on behalf of a
  company, you can now also provide the branch number.
* [:backend:`5216`] Skipped sending the payment update email when the payment update emailaddress is
  empty and the flag "Wait for payment to register" is enabled, so now a user won't receive

>   multiple emails.

Reviewer: ```suggestion
* [:backend:`5216`] The payment update email is now no longer sent when the payment update emailaddress is
  empty and the flag "Wait for payment to register" is enabled to prevent the backoffice receiving duplicate
  emails.
```

* [:backend:`5887`] Added ``gor.openbareRuimteNaam`` to verblijfsadres and bezoekadres in StUF-ZDS.
* [:backend:`5857`] Changed the default to unchecked for the "Include deceased" checkbox in the

>   family members prefill configuration.

Reviewer: ```suggestion
* [:backend:`5857`] In the family members prefill configuration, deceased family members are now excluded by default.
```

* [:backend:`4746`] Added environment to the subject of the daily report email.

> * [:backend:`5977`] Ensured that logic actions can't change the "disabled" property for layout fields.

Reviewer: ```suggestion
* [:backend:`5977`] Ensured that logic actions can't change the "disabled" property for components that don't support it.
```

Security fix
------------

* [:cve:`CVE-2026-28803`] Fixed a vulnerability where attackers could view form
  submission data. See :ghsa:`GHSA-2g49-rfm6-5qj5` for details and instructions on how

>   to detect possible intrusions. This advisory will be published on Wednesday March 11th.

Reviewer: ```suggestion
* [:cve:`CVE-2026-28803`] Fixed a vulnerability where attackers could view form
  submission data. See :ghsa:`GHSA-2g49-rfm6-5qj5` for details and instructions on how
  to detect possible intrusions.
```

* [:backend:`5950`] Fixed BAG-error responses being cached.
* [:backend:`6040`] Fixed fields inside editgrids unexpectedly being cleared.
* [:backend:`6046`] Fixed the result of a variable action being overwritten by an
  untriggered logic action on the same component that affects its visibility.
* [:backend:`6028`] Fixed a regression causing fields to skip final validation if
  they're made visible through backend logic.
* [:backend:`6045`] Fixed infinite loop in logic check with the new renderer when a
  repeating group with ``clearOnHide: false`` is hidden while a field inside it has
  ``clearOnHide: true``.
* [:backend:`5685`] Fixed infinite loop in logic check due to already submitted
  repeating group data not being used in logic evaluation.
* [:backend:`5967`] Fixed the theme-specific "back to main website"-link and favicon not
  being applied correctly.
* Validation errors in the cosign lookup flow are now displayed with proper error styling.
* [:backend:`6068`] Fixed attachement possibly not being uploaded when component's visiblity
  is affected by logic rule.
* [:backend:`5701`] Fixed ``selectboxes`` component, so it won't crash when removed if the

>   Objects API v1 is used.

Reviewer: ```suggestion
* [:backend:`5701`] Fixed ``selectboxes`` component, so it won't crash the admin UI when removed if the
  Objects API v1 is used.
```

> * [:backend:`5864`] Stopped showing warnings in the form history page when the application version is

Reviewer: ```suggestion
* [:backend:`5864`] Fixed the warnings display in the form history page when the application version is
```

  the current one.
* [:backend:`6110`] Fixed broken asset cache by configuring staticfiles storage entry.
* [:backend:`5938`] Fixed incorrect DigiD error message being shown when the user
  cancels a DigiD login via the "DigiD via OIDC" plugin.
* [:backend:`6140`] Fixed a crash when conditional logic is used inside a fieldset
  inside a repeating group.

> * [:backend:`5924`] Fixed valdation messages showing the wrong (non-applicable) steps.

Reviewer: ```suggestion
* [:backend:`5924`] Fixed validation messages being linked to the wrong (non-applicable) steps, causing
  confusing error mesages in the public frontend.
```

* Fixed admin-index border not stretching full width.

Project maintenance
-------------------

* [:backend:`5879`] Extended the component problem detection and added automatic fixes
  for the issues.
* Cleaned up inconsistent test setups.
* Updated ``pyright`` and improved type checking.
* Replaced the unmaintained bleach with nh3 HTML-sanitizer library.
* Upgraded to storybook 9.
* Updated backend dependencies to their latest versions:

  - protobuf
  - weasyprint
  - django
  - urllib3
  - cbor2

>   - lodash

Reviewer: is a frontend dependency :-)

Author: ugh I decided it deserved mentioning both in backend and frontend. Removed

  - cryptography
  - Pillow

>   - webpack

Reviewer: is a frontend dependency :-)

Author: Moved where it belongs

We'll work on bridging this gap, possibly through an upgrade to Django 6.0.

**Textfield Location tab**

> Deriving the location via text fields is depricated if favour of AdressNL component.

Reviewer: ```suggestion
Deriving the location via text fields is deprecated in favour of the AdressNL component.
```

As mentioned in the release notes of Open Forms 3.4, the new renderer doesn't support
autofilling address on the Textfield component.
We plan to resolve all outstanding address autofill issues in the AddressNL component before
removing the old renderer and Textfield address autofill support.

**Legacy logic evaluation**

Old logic evaluation will be removed after the new logic evaluation is considered stable.

> We plan to extensively test logic rules before retiring the old logic engine.

Reviewer: ```suggestion
**Legacy logic evaluation**

The old logic evaluation will be removed in Open Forms 4.0. We will prioritize bugfixes in the new 
logic evaluation to make sure this migration will be as smooth as possible.
```
in your version it creates the impression that the new logic evaluation is not yet stable or ready for production use - we definitely intend it to be ready enough.

**File Registration tab**

We plan to replace the documenttype configuration by URL in favour of catalog + documenttype

> description and move the configuration to the registration plugin options.

Reviewer: ```suggestion
**File Registration tab**

We plan to replace the documenttype configuration by URL in favour of catalog + documenttype
description and move the configuration to the registration plugin options. We're exploring automatic
migration options.
```

**Elastic APM**

Elastic APM agent integration is deprecated in favour of Open Telemetry. We plan to remove
it in Open Forms 4.0.

**ZGW APIs registration**

The process of creating a zaak and using its identifier as the public reference for a submission is deprecated.
We plan to change the default for "Use generated zaaknummer" to unchecked in an Open Forms 4.0.

**StUF-ZDS registration**

The usage of the branch number retrieved from the form's submission will be replaced in Open Forms 4.0

> by prioritizing the provided branch number after login.

Reviewer: ```suggestion
**StUF-ZDS registration**

The usage of the branch number retrieved from the form's submission data will be replaced in 
Open Forms 4.0 by favouring the branch number obtained through the eHerkenning login.
```

**DigiD/eHerkenning OIDC Strict mode**

> The default SameSite cookie mode for DigiD/eHerkenning via OIDC authentication is changing from Lax to Strict.

Reviewer: ```suggestion
**DigiD/eHerkenning OIDC Strict mode**

We will make Strict mode the default (currently it's Lax mode) for DigiD and eHerkenning login via OIDC. This
means that the identity provider is required to provide properties like the level of assurance, acting subject identifier and service IDs when mandates (DigiD, eHerkenning) are applicable.
```

**Haal Centraal BRP Personen bevragen 1.3**

We are dropping for Haal Centraal BRP Personen bevragen < 2.0. This version was never used in production
and only added testing and maintenance overhead without any benefit.

> We plan to remove it fully in Open Forms 4.0.

Reviewer: ```suggestion
**Haal Centraal BRP Personen bevragen 1.3**

We are dropping support for Haal Centraal BRP Personen bevragen 1.3. In practice, only version 2.0 has
been taken into production (nation-wide). In Open Forms 4.0 it will be removed entirely.
```

**``clearOnHide`` behaviour change**

When a component is hidden and cleared, the associated variable will now be removed entirely rather

> than reset to its default value. This may affect existing logic rules that depend on the previous behaviour.

Reviewer: ```suggestion
**"Clear on hide" logic behaviour changes**

When a component is hidden and cleared, the associated variable will be removed entirely rather
than reset to its default (or empty) value. This may affect existing logic rules that (unknowingly) depend on the previous behaviour. We're exploring tools to help detect these situations so that you can update your forms.
```

**Extensions**

Extensions do not seem to be created commonly by third parties and may be removed in
the future. The public interfaces for plugins/extensions will become semi-private API allowing us

> to make potentially breaking changes without making a new major version of OF.

Reviewer: ```suggestion
to make potentially breaking changes without bumping Open Forms to a new major version.
```

> We plan to remove official extension support and move existing extensions into the core.

Reviewer: ```suggestion
```
we're not planning to move them into core, just don't commit to stable public interfaces any longer for interfaces :)

Author: Oh I think I've read something like this in the deprecation issue and just copied it

Reviewer: probably some unresolved thoughts that I dumped!
